assignments/burmanm/week-3.scm

```diff
@@ -0,0 +1,279 @@
+(fixed-point (lambda (x) (average x (/ (log 1000) (log x)))) 2.0)
+
+;; We get 9 iterations to reach 4.555537551999825
+;; If we select different initial guess, then it may take different amount of iterations (such as 10.0 requiring 10 iterations)
+
+;; Exercise 1.37
+
+(define (count-fract-recursive n d k)
+  (define (frac-r i)
+    (if (< i k)
+	(/ (n i) (+ (d i) (frac-r (+ i 1))))
+	(/ (n i) (d i))))
+  (frac-r 1))
+
+;; 11 iterations is enough to get 4 decimal places
+
+(define (cont-fract n d k)
+   (define (frac-iter i sum)
+       (if (= i 0)
+           sum
+           (frac-iter (- i 1) (/ (n i) (+ (d i) sum)))))
+   (frac-iter (- k 1) (/ (n k) (d k))))
+
+;; Exercise 1.38
+;; d(i) = 2(i + 1) / 3
+
+(define (d i)
+   (if (= 0 (remainder (+ i 1) 3))
+       (/ (* 2 (+ i 1)) 3)
+       1))
+
+(define (euler-number k)
+   (+ 2 (cont-fract (lambda (i) 1) d k)))
+
+(euler-number 10)
+;; Value: 2721/1001
+
+
+;; Exercise 1.39
+
+(define (tan-cf x k)
+  (cont-fract
+   (lambda (i)
+     (if (= i 1)
+	 x
+	 (- (square x))))
+   (lambda (i) (- (* 2 i) 1))
+   k))
+
+;; Exercise 1.40
+
+(define (cubic a b c)
+  (lambda (x) (+ (* x x x) (* a (* x x)) (* b x) c)))
+
+;; Exercise 1.41
+
+(define (double f)
+  (lambda (x) (f (f x))))
+
+;; scheme@(guile-user)> (define (inc x) (+ x 1))
+;; scheme@(guile-user)> ((double inc) 1)
+
+;; Exercise 1.42
+
+(define (compose f g)
+  (lambda (x) (f (g x))))
+
+;; scheme@(guile-user)> ((compose square inc) 6)
+;; $3 = 49
+
+;; Exercise 1.43
+
+(define (repeated f n)
+  (if (= n 1)
+      f
+      (repeated (lambda (x) (f (f x))) (- n 1))))
+
+;; Could use compose from above instead of repeating..
+
+;; scheme@(guile-user)> ((repeated square 2) 5)
+;; $6 = 625
+
+;; Exercise 1.44
+
+(define dx 0.00001)
+
+(define (smooth f)
+  (lambda (x) (/ (+ (f (+ x dx)) (f x) (f (- x dx))) 3)))
+
+(define (nfold-smooth f n)
+  ((repeated smooth n) f dx))
+
+;; Exercise 1.45
+
+;; ...
+
+;; Exercise 1.46
+;; From old exercices..
+(define (improve guess x)
+  (average guess (/ x guess)))
+
+(define (good-enough? guess x)
+  (< (abs (- (square guess) x)) 0.001))
+
+(define (average x y)
+  (/ (+ x y) 2))
+
+(define (sqrt-iter guess x)
+  (if (good-enough? guess x)
+      guess
+      (sqrt-iter (improve guess x) x)))
+
+;; Redo the same sqrt-iter with generic implementation
+
+(define (iterative-improve gn ig)
+  (lambda (x)
+    (define (iter-guess guess)
+      (if (gn guess x)
+	  guess
+	  (iter-guess (ig guess x))))
+    (iter-guess x)))
+
+(define (sqrt x)
+  ((iterative-improve good-enough? improve) x))
+
+
+;; From old fixed-point
+(define tolerance 0.00001)
+(define (fixed-point f first-guess)
+   (define (close-enough? v1 v2)
+     (< (abs (- v1 v2)) tolerance))
+   (define (try guess)
+     (let ((next (f guess)))
+       (if (close-enough? guess next)
+           next
+           (try next))))
+   (try first-guess))
+
+;; Redo .. and something's not working (wrong number of arguments.. to what?)
```

Comment:
In iterative-improve, L252, it uses (gn guess x), but L277, the gn procedure only takes one argument.